### PR TITLE
Enable moderating spams in comments

### LINKED ISSUE #1341: unable to remove spam from comments listing
using Wordpress for Android v2.8 on a Galaxy S4.
Self-hosted Wordpress

In my listing of comments on my site, there are some 20 spam comments, listed as "Pending". However, trying to remove them fails, and I will explain why.

I had them listed in the app, but then went to the blog with my browser and removed them in Wordpress instead. Ever since, I cannot remove them from the app. When I try to either mark them spam or trash them, I get "an error occurred while moderating" and the comment is not removed from the listing. The log shows that there is an error 404 on the request.

### PR DESCRIPTION
**Pull request by fadils**
Fix #1341

### CODE REVIEW

**Comment by fadils** [comment]
Basically this PR is doing two things. From my observation, comment that is moderated as spam FROM the website, then these comments won't be included in `wp.getComments` general result. That's why we're seeing that it's not updated in the app. So, first, we will additionally call `wp.getComments` with specific spam status to include the spam-moderated comments.

Second, the "error while moderating" happens because the user is trying to do something to a file that is already deleted in the server (it doesn't exist anymore). In other words, localdb is NOT synchronized with its remote counterpart. So, we will synchronize the localdb so that invalid/deleted comments will not show up in the app.

**Comment by daniloercoli** [comment]
<blockquote>
Basically this PR is doing two things. From my observation, comment that is moderated as spam FROM the website, then these comments won't be included in wp.getComments general result. That's why we're seeing that it's not updated in the app. So, first, we will additionally call wp.getComments with specific spam status to include the spam-moderated comments.
</blockquote>
Can we just reload comments from the server by calling the `wp.getComments` in the "normal" way `result = (Object[]) client.call("wp.getComments", commentParams);` without requesting SPAM comments, and just remove all of the old comments from the DB when the server replies with the new comments lists?

In other words, I wonder why we need to request spam comments from the server, and then remove them later in a for loop. Isn't easier to remove all of the comments that are stored in the DB for the current blog, and add comments that arrives from the server?

**Comment by fadils** [comment]
> Can we just reload comments from the server by calling the wp.getComments in the "normal" way result = (Object[]) client.call("wp.getComments", commentParams); without requesting SPAM comments, and just remove all of the old comments from the DB when the server replies with the new comments lists?

Yes, we can do that. However, comments moderated as SPAM won't be shown at all in the app.
I'm not sure if we want to do this, since in 

https://github.com/wordpress-mobile/WordPress-Android/blob/36ef8fe9c8c63948f33f039847d47237a16e2dc4/WordPress/src/main/java/org/wordpress/android/ui/comments/CommentAdapter.java#L242

the current one sets that spam should be shown instead.

> In other words, I wonder why we need to request spam comments from the server, and then remove them later in a for loop. Isn't easier to remove all of the comments that are stored in the DB for the current blog, and add comments that arrives from the server?

Completely my bad. I mixed up with how post synchronization is done. Thank you for this. :relieved: 

As an extra note, if we want to go the route with showing the spam, will this PR also fix #60?

**Comment by fadils** [comment]
Db synchronization is updated, thanks to @daniloercoli. Could you please let me know how we want to do with showing the spams?
If we're going to NOT showing the spam, then I'll update this PR accordingly.

**Comment by daniloercoli** [comment]
<blockquote>Could you please let me know how we want to do with showing the spams?</blockquote>
@roundhill - What we want to do with SPAM comments...should we download them when a refresh of comments is started? IIRC, long long time ago we decided to not show up spam comments in the list to not fill the comments screen with spams, maybe it's the time to rethink that.

What if we add a entry in the option menu, that gives the user the ability to hide/show comments? In that way we always synch "all" comments, but the adapter shows the spam status comments only when the user has selected that option.

**Comment by daniloercoli** [comment]
@fadils - I see a problem here https://github.com/wordpress-mobile/WordPress-Android/pull/2585/files#diff-d530f0ee974683d1caa9454406b4d695R397 in the method that does the synchronization of SPAM comments. I see 2 remote calls there, the first one that downloads the first 20 comments (without downloading SPAM comments), and the second call that does the synchronization of SPAM comments by downloading the 20 most recent spam comments. 

- If you use 2 different calls there, you are not sure you're downloading the most recent 40 comments...
- Also, what's happens when the user reached the bottom of comments screen, and the "load more" comments is called?

Could you try to download ALL comments by using the snippet below? 

` HashMap<String, Object> withSpam = new HashMap<>();
        withSpam.put("status", "all");
        Object[] commentParamsWithSpam = {blog.getRemoteBlogId(), blog.getUsername(),
                blog.getPassword(), withSpam};
`

If you pass the status `all` the server should reply with all comments.

**Comment by fadils** [comment]
> If you pass the status all the server should reply with all comments.

So much better. Will change it now.

**Comment by fadils** [comment]
Confirmed. Passing `("status", "all")` doesn't return the SPAM. :confused:

**Comment by daniloercoli** [comment]
Try by using `any`.

**Comment by daniloercoli** [comment]
If calling getComments with `("status", "any")` doesn't return SPAM comment, you can try with a comma-separated string.
`("status", "all,spam")`  or `("status", "hold,approve,spam")`

**Comment by fadils** [comment]
Last commit also handles how "load more" is conducted as well as make sure that comment that is moderated from website, and appears in the list after "load more", gets updated too.

**Comment by daniloercoli** [comment]
Maybe it's my fault, but the logic that does the synch to the DB seems pretty complex, and can lead to errors. 

What if we follow these simple rules:
- If you're requesting the first page of comment (when the parameter offset is not present: first time you open the comments screen, or when you do a PTR)  then you can delete all the comments in the DB for that blog and add those returned from the server.

- If we're in a loading more step (parameter offsert available): delete all the comments for that blog that are beyond the offset. Then add all the comments returned from the server. The first part of this second step (sorry for the pun) should not be necessary, since we should already have deleted all the old comments at this point. Anyway keep it for safety.

Before changing your implementation is better to hear what @roundhill thinks. I bet he has better ideas on how to handle this correctly.

**Comment by fadils** [comment]
@daniloercoli if I understand you correctly, would that work if we retrieve all comments from the server simultaneously?
I tried it yesterday. I just felt the load time a bit longer (although haven't really put a timer on it :crying_cat_face: ) 
Would be awesome to hear what @roundhill has to say.

**Comment by roundhill** [comment]
I think @daniloercoli has the right idea with those rules. If a pull-to-refresh is initiated (and we get a response), we should essentially start over by wiping out the comments table and using the new data only.

@nbradbury might have some ideas as well, since he refactored the comments area a while ago.

**Comment by nbradbury** [comment]
> @nbradbury might have some ideas as well, since he refactored the comments area a while ago.

Although wiping the data and getting it fresh has its downsides, I think you'll find it to be a much simpler and more reliable approach. Keeping the comments cached runs the risk of "gaps" - ie: missing comments between the ones last downloaded and the ones you just downloaded (a source of frustration with the reader).

**Comment by fadils** [comment]
Thanks @daniloercoli @roundhill @nbradbury 
:clap: :clap: :clap:

**Comment by daniloercoli** [comment]
@fadils - Is there a chance to rewrite the function that deletes comments?
I wonder why there a select, and then multiple calls to `CommentTable.deleteComment(localBlogId, comment.commentID);` 
Could it be done in one single DELETE command?

**Comment by fadils** [comment]
@daniloercoli, I'm not sure.
The select is, IMHO, important to facilitate the two rules previously stated.

One way for single DELETE command is to be used when `offset` isn't present (-1). This happens when PTR is initiated (Rule 1). Then, we probably can use `CommentTable.reset`.
But, when the `loadMore` operation is initiated (Rule 2), then I thought we need to specify certain comments to delete, then traverse them from the cursor. 

Let me know what you have in mind. Then I can update this PR right away. =)

**Comment by daniloercoli** [comment]
<blockquote>One way for single DELETE command is to be used when offset isn't present (-1). This happens when PTR is initiated (Rule 1)</blockquote>
Yes, would be nice to have a function called `deleteCommentsForBlog(int blog)` that does erase all the comment in one single transaction. 
In the current implementation, when I do a PTR, comments are deleted in a for-loop, but I can have scrolled far-away down in comments, and have 200+ comments in the table. PTR is ofter the most common user case. You open the comment screen and refresh them.

**Comment by fadils** [comment]
@daniloercoli sorry for long response. Just had the chance to go back to Wordpress project today.
Actually, there's already a `deleteCommentsForBlog(int blog)` function, which in turn execute a DELETE statement with blog's id as its WHERE clause. I'm not sure if the function will perform faster on a different level, compared to previous for-loop.

In any case, I already updated the sync as requested. :grin:

Let me know if now is good.

File: WordPress/src/main/java/org/wordpress/android/ui/comments/CommentsListFragment.java [hunk 1]
```diff
@@ -469,6 +469,9 @@ protected CommentList doInBackground(Void... args) {
                 int numExisting = getCommentAdapter().getCount();
                 hPost.put("offset", numExisting);
                 hPost.put("number", COMMENTS_PER_PAGE);
+            } else if (!mCanLoadMoreComments ||
```
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/ui/comments/CommentsListFragment.java]: Is it possible to remove this part of the if condition?
If I have 200+ comments in the list, and do a PTR the app asks for 200+ comments to the server, that in some self hosted installation can lead to OOM errors. If the user launches a PTR the default number of comments should be retrieved from the server.

**Comment by daniloercoli** [comment]
The code is OK, but enabling SPAM in comments list could result in a sub-optimal UX. 
Ref: https://github.com/wordpress-mobile/WordPress-Android/issues/58

(On a real site you'll see a long list of SPAM comments at the beginning of the list, making hard to approve legit comments. )

<b>Do not merge</b> this PR until #58 is fixed.
